test/test_torch.py

```diff
@@ -4427,6 +4427,122 @@ def test_diff(self, device, dtype):
 
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.long, torch.float32, torch.complex64)
+    def test_gradient_tensor_list(self, device, dtype):
+        inputs = (
+            ((4, 4, 4), [(1., 3.0, 6., 10.), (1., 3.0, 4., 90.), (1., 2.5, 3.4, 7.)], [0, 1, 2], 1),
+            ((4, 4, 4, 3), [(1., 3.0, 6., 10.), (1., 2.5, 7.)], [0, -1], 1),
+            ((4, 5, 2, 4, 3), [(1., 3.0, 6., 10.), (1., 3.4, 7.)], [-2, -1], 1),
+            ((4, 3, 6, 5, 3), [(1., 3.0, 6., 10.), (2.5, 3.4, 7.)], [0, 1], 1),
+            ((4, 3, 2, 5, 3), [(1., 3.0), (3.4, 7., 6.0)], [2, 1], 1),
+        )
+
+        for size, coordinates, dim, edge_order in inputs:
+            for contig in [True, False]:
+                t = make_tensor(size, device, dtype, noncontiguous=not contig)
+                t_n = t.cpu().numpy()
+                coordinates_list = []
+                for space in coordinates:
+                    if dtype == torch.complex64 and isinstance(space, float) and abs(space) < 1e-7:
+                        return
+                    coordinates_list.append(torch.tensor(space, dtype=float, device=device))
+                result = torch.gradient(t, domain=coordinates_list, dim=dim, edge_order=edge_order)
+                result_n = np.gradient(t_n, *coordinates, axis=dim, edge_order=edge_order)
+                if not isinstance(result_n, list):
+                    result_n = [result_n]
+                self.assertEqual(result, result_n)
+
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.long, torch.float32, torch.complex64)
+    def test_gradient_float(self, device, dtype):
+        inputs = (
+            ((5,), [2.73], 0, 1),
+            ((5, 5), 2.0, 0, 1),
+            ((2, 2, 2), [1.17, 3.34, 1.73], [0, 1, 2], 1),
+            ((4, 4, 4), 1.7 , [0, 1, 2], 1),
+            ((4, 3, 4), 1.7 , [-2, -1], 1),
+            ((4, 3, 4, 5), [1.7, 3.93], [0, -1], 1),
+            ((4, 3, 4, 5), 1.7 , [1, 2], 1),
+            ((4, 3, 4, 5), 1.7 , [0, 1], 1),
+            ((4, 3, 4, 5), 1.7 , [2, -1], 1),
+            ((4, 3, 4, 5), -1.7 , [2, -1], 1),
+            ((4, 3, 4, 5), 0.0 , [2, -1], 1),
+        )
+
+        for size, spacing, dim, edge_order in inputs:
+            if dtype == torch.complex64 and isinstance(spacing, float) and abs(spacing) < 1e-7:
+                return
+            for contig in [True, False]:
+                t = make_tensor(size, device, dtype, noncontiguous=not contig)
+                t_n = t.cpu().numpy()
+                result = list(torch.gradient(t, domain=spacing, dim=dim, edge_order=edge_order))
+                spacing_ = spacing if isinstance(spacing, list) else [spacing]
+                result_n = np.gradient(t_n, *spacing_, axis=dim, edge_order=edge_order)
+                if not isinstance(result_n, list):
+                    result_n = [result_n]
+                for i in range(len(result)):
+                    if dtype == torch.complex64 and (result[i].isinf().any() or result[i].isnan().any()):
```

Comment:
add a comment for preprocessing

maybe this function should be a comparison helper function that runs whenever comparing with NumPy?